### PR TITLE
Fix listing of Scannable URLs and re-fetching upon reaching Done step of Onboarding Wizard

### PR DESCRIPTION
**Pull request by westonruter**
## Summary

Originally discovered at https://github.com/ampproject/amp-wp/pull/6773#issuecomment-989219662.

In Reading Settings when I configure WordPress to show a specific page for the homepage and posts page:

image

The result is at the moment _no_ homepage entry being added to the list of scannable URLs:

https://github.com/ampproject/amp-wp/blob/4cd08b2982682b5641c8376694f5f5ad414ee636/src/Validation/ScannableURLProvider.php#L110-L119

This results in the review screen of the Onboarding Wizard not including the homepage. This is a bug with the `ScannableURLProvider`. It should be including here two entries for Homepage and Blog. If I instead select to show the Latest Posts on the homepage, then it does include the Homepage entry. The issue is that upon finishing the Onboarding Wizard, the Scannable URLs are no longer being re-fetched according to the latest saved settings. This is resulting in inappropriate URLs being listed, such as for legacy Reader mode:

> image

Regardless of what I have selected for the front page in Reading settings, I'm seeing non-singular templates in the Done screen when I have selected legacy Reader mode. In contrast, only Post, Page, and Media should be shown:

> image

## Checklist

- [ ] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [ ] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

### CODE REVIEW

File: src/Validation/ScannableURLProvider.php [hunk 1]
```diff
@@ -110,12 +110,29 @@ public function get_urls() {
...
+					'label' => __( 'Homepage', 'amp' ),
+				];
+			}
+			if ( $this->is_template_supported( 'is_home' ) && get_option( 'page_for_posts' ) ) {
+				$urls[] = [
+					'url'   => get_permalink( get_option( 'page_for_posts' ) ),
+					'type'  => 'is_home',
+					'label' => __( 'Blog', 'amp' ),
+				];
+			}
```
Comment (westonruter) [inline src/Validation/ScannableURLProvider.php]: When in legacy Reader mode, these should only be included only if AMP has been explicitly enabled for the page per this logic in `amp_is_available()`:

https://github.com/ampproject/amp-wp/blob/4cd08b2982682b5641c8376694f5f5ad414ee636/includes/amp-helper-functions.php#L541-L549

**Comment by delawski** [comment]
I think not refetching scannable URLs after options are saved is a regression. I think it was working like that before. Anyway, in 11f8ce5 this has been fixed.

What's more, I've noticed that the plugin suppression data has been refetched after every scan while it only is needed on the AMP Settings screen (where the plugin suppression can be changed). In 56f0d8b the plugin suppression will only be refetched on the settings screen:

| AMP Settings<br>Plugin suppression data is refetched | Onboarding Wizard<br>Suppression data is no longer refetched after scan |
| --- | --- |
| Screenshot 2021-12-09 at 15 37 05 | Screenshot 2021-12-09 at 15 37 44 |

File: src/Validation/ScannableURLProvider.php (hunk 1 above)
Comment (westonruter) [reply src/Validation/ScannableURLProvider.php]: I'm going to add some tests for this.

**Comment by westonruter** [comment]
> I think not refetching scannable URLs after options are saved is a regression. I think it was working like that before. Anyway, in [11f8ce5](https://github.com/ampproject/amp-wp/commit/11f8ce5d764be0b23c0ecec580a3ba662a9aaa7d) this has been fixed.

@delawski I'm noticing one additional case that doesn't seem to be accounted for. If I go through the wizard to completion after having selected legacy Reader for example, and then I click Previous to try changing to Transitional (or Standard or a Reader theme), when I go to finish _again_ the scannable URLs are being shown only reflecting the state after the _first_ save. It needs to re-fetch after _every_ save.


https://user-images.githubusercontent.com/134745/145531612-3c6df996-a30b-488d-a6eb-98c266317405.mov

(I reloaded the wizard halfway through this video.)

File: src/Validation/ScannableURLProvider.php (hunk 1 above)
Comment (westonruter) [reply src/Validation/ScannableURLProvider.php]: Done in 59db0ed35

**Comment by delawski** [comment]
> It needs to re-fetch after _every_ save.

Good catch. This is now fixed in 4b6ddab:

https://user-images.githubusercontent.com/478735/145574227-76e4b7e0-482f-4517-8f96-2355a3deb21d.mp4

**Comment by delawski** [comment]
Even though there are a few E2E tests failing, I think that the PR is ready for review (I'll take care of the tests as soon as I'm done with other task).

**Comment by delawski** [comment]
Reverting to draft as I noticed some issue when changing a previewed page. On it now.

**Comment by delawski** [comment]
E2E tests were failing for a good reason. The Site Scanner could get stuck in an incorrect state in the Onboarding Wizard after some recent updates. I think 19db5fa should fix the issue, let's see what will be the action output.

**Review by delawski — APPROVED** [review]
Everything works as expected after reverting 3f3cb57. I'm approving but since I've written quite a bit of code in this PR, it'd be great to have another pair of eyes do a review, too.

**Comment by dhaval-parekh** [comment]
✅  `ScannableURLProvider` now includes entries for the front-page and blog page if it is set in "Reading" settings.

✅  On "Onboarding Wizard" It refetching Scannable URLs after saving the configuration and reflecting it to the Preview screen.

⚠️  However, It's fetching twice.

https://user-images.githubusercontent.com/8168027/145806095-a225c32d-3581-46fb-a463-74d3e1f3697b.mov


Other than this all look good to me.

**Comment by delawski** [comment]
> ⚠️ However, It's fetching twice.

I can replicate it locally. I'll have a look into this shortly.

**Comment by delawski** [comment]
@dhaval-parekh With cdc4a7d the `scannableUrls` should be fetched only once. Please, recheck.

**Comment by dhaval-parekh** [comment]
> With [cdc4a7d](https://github.com/ampproject/amp-wp/commit/cdc4a7da03c8bcaaeca2d2f6b69d1930e5f40126) the `scannableUrls` should be fetched only once. Please, recheck.

Thanks, Piotr I can see that It's fetching scannable URL only once. and with that, I can mark this issue as **QA Passed**

**Review by dhaval-parekh — APPROVED** [review]

File: src/Validation/ScannableURLProvider.php [hunk 2]
```diff
@@ -71,8 +71,12 @@ public function get_supportable_templates() {
 				'is_singular',
 			];
 			if ( 'page' === get_option( 'show_on_front' ) ) {
-				$allowed_templates[] = 'is_home';
-				$allowed_templates[] = 'is_front_page';
+				if ( get_option( 'page_for_posts' ) && amp_is_post_supported( get_option( 'page_for_posts' ) ) ) {
+					$allowed_templates[] = 'is_home';
+				}
+				if ( get_option( 'page_on_front' ) && amp_is_post_supported( get_option( 'page_for_posts' ) ) ) {
```
Comment (westonruter) [inline src/Validation/ScannableURLProvider.php]: Oops. This should have been:

```suggestion
				if ( get_option( 'page_on_front' ) && amp_is_post_supported( get_option( 'page_on_front' ) ) ) {
```

**Pull request merged**